[Bazel] Define BUILTIN_THREAD_POINTER where appropriate on x86-64 linux

boomanaiden154 opened the pull request:
This patch defines BUILTIN_THREAD_POINTER in config.bzl when appropriate on x86-64 linux only. This is needed to build exegesis properly as certain functionality breaks if this is not enabled. This option is only supported on relatively recent compiler versions, but given most people using the bazel build are using very recent toolchains, this shouldn't be an issue.

boomanaiden154 committed:
[Bazel] Define BUILTIN_THREAD_POINTER where appropriate on x86-64 linux

This patch defines BUILTIN_THREAD_POINTER in config.bzl when appropriate
on x86-64 linux only. This is needed to build exegesis properly as certain
functionality breaks if this is not enabled. This option is only
supported on relatively recent compiler versions, but given most people
using the bazel build are using very recent toolchains, this shouldn't
be an issue.

rnk commented:
I guess in general this does require a configure check, since GCC __has_builtin returns true for this, but then errors out during code generation:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96952#c2

We should have to replicate knowledge of which targets GCC supports this builtin on in order to forego the configure check.

MaskRay commented:
The CMake check was introduced in 822c31a0fe827a6f33601ad6b44eb2818350da36 and adjusted (`used`) in a54545ba6514802178cf7cf1c1dd9f7efbf3cde7. Perhaps we can also disable Bazel builds if GCC<=10 or Clang<=13 at another layer?
People who use this unofficial build system seem to use very new Clang or GCC.
(Also we don't have build bots for older compilers, so  they may not even work.)

boomanaiden154 commented:
> The CMake check was introduced in [822c31a](https://github.com/llvm/llvm-project/commit/822c31a0fe827a6f33601ad6b44eb2818350da36) and adjusted (`used`) in [a54545b](https://github.com/llvm/llvm-project/commit/a54545ba6514802178cf7cf1c1dd9f7efbf3cde7). Perhaps we can also disable Bazel builds if GCC<=10 or Clang<=13 at another layer? 

Didn't realize it was modified further after landing. Others would have to chime in on how to disable Bazel builds with old toolchains/whether or not that is even wanted.

>People who use this unofficial build system seem to use very new Clang or GCC. (Also we don't have build bots for older compilers, so they may not even work.)

I figured most people using Bazel would be using a recent toolchain, but wasn't entirely sure and would rather be a little bit more careful, especially as the exact toolchain requirements aren't documented anywhere, so this could theoretically cause breakage on platforms that weren't explicitly marked as unsupported.

> We probably don't even need to care about platforms where GCC doesn't define `__builtin_thread_pointer`: people don't use Bazel on these platforms.

If that's the case, then I can just add it to the Linux x86-64 defines in `./utils/bazel/llvm-project-overlay/llvm/config.bzl`. I guess gcc 10 is >3 years old at this point.

boomanaiden154 committed:
Add comment

boomanaiden154 commented:
> I like that solution.

Thanks for working with me to find a solution. I've implemented that and added a comment explaining the details. Please let me know if there are any refactorings/style issues with how I've done it as I'm not exactly sure what the canonical Bazel way to do this would be.

boomanaiden154 commented:
Bump on this when reviewers have a chance. Thanks!

MaskRay commented on utils/bazel/llvm-project-overlay/llvm/config.bzl:
```diff
@@ -94,7 +94,10 @@ llvm_config_defines = os_defines + select({
     "@bazel_tools//src/conditions:linux_aarch64": native_arch_defines("AArch64", "aarch64-unknown-linux-gnu"),
     "@bazel_tools//src/conditions:linux_ppc64le": native_arch_defines("PowerPC", "powerpc64le-unknown-linux-gnu"),
     "@bazel_tools//src/conditions:linux_s390x": native_arch_defines("SystemZ", "systemz-unknown-linux_gnu"),
-    "//conditions:default": native_arch_defines("X86", "x86_64-unknown-linux-gnu"),
```
We can define HAVE_BUILTIN_THREAD_POINTER for all targets except `powerpc64le-linux-gnu-gcc`.

`clang --target=powerpc64le` supports `__builtin_thread_pointer`

boomanaiden154 replied:
I've refactored and set it to apply to all platforms except for `linux_ppc64le`. Does everything still look good to you?

boomanaiden154 committed:
Refactor, add on all Linux platforms except for ppc64
